nameresolution/mdns/mdns.go

```diff
@@ -384,81 +673,89 @@ func (m *resolver) refreshAllApps(ctx context.Context) error {
 }
 
 // browse will perform a non-blocking mdns network browse for the provided app id.
-func (m *resolver) browse(ctx context.Context, appID string, onEach func(ip string)) error {
+func (m *Resolver) browse(ctx context.Context, appID string, onEach func(ip string)) error {
 	resolver, err := zeroconf.NewResolver(nil)
 	if err != nil {
-		return fmt.Errorf("failed to initialize resolver: %e", err)
+		return fmt.Errorf("failed to initialize resolver: %w", err)
 	}
 	entries := make(chan *zeroconf.ServiceEntry)
 
-	// handle each service entry returned from the mDNS browse.
-	go func(results <-chan *zeroconf.ServiceEntry) {
-		for {
-			select {
-			case <-ctx.Done():
-				if errors.Is(ctx.Err(), context.Canceled) {
-					m.logger.Debugf("mDNS browse for app id %s canceled.", appID)
-				} else if errors.Is(ctx.Err(), context.DeadlineExceeded) {
-					m.logger.Debugf("mDNS browse for app id %s timed out.", appID)
-				}
+	handleEntry := func(entry *zeroconf.ServiceEntry) {
+		for _, text := range entry.Text {
+			if text != appID {
+				m.logger.Debugf("mDNS response doesn't match app id %s, skipping.", appID)
 
-				return
-			case entry := <-results:
-				if entry == nil {
-					break
-				}
+				break
+			}
 
-				for _, text := range entry.Text {
-					if text != appID {
-						m.logger.Debugf("mDNS response doesn't match app id %s, skipping.", appID)
+			m.logger.Debugf("mDNS response for app id %s received.", appID)
 
-						break
-					}
+			hasIPv4Address := len(entry.AddrIPv4) > 0
+			hasIPv6Address := len(entry.AddrIPv6) > 0
 
-					m.logger.Debugf("mDNS response for app id %s received.", appID)
+			if !hasIPv4Address && !hasIPv6Address {
+				m.logger.Debugf("mDNS response for app id %s doesn't contain any IPv4 or IPv6 addresses, skipping.", appID)
 
-					hasIPv4Address := len(entry.AddrIPv4) > 0
-					hasIPv6Address := len(entry.AddrIPv6) > 0
+				break
+			}
 
-					if !hasIPv4Address && !hasIPv6Address {
-						m.logger.Debugf("mDNS response for app id %s doesn't contain any IPv4 or IPv6 addresses, skipping.", appID)
+			var addr string
+			port := entry.Port
 
-						break
-					}
+			// TODO: we currently only use the first IPv4 and IPv6 address.
+			// We should understand the cases in which additional addresses
+			// are returned and whether we need to support them.
+			if hasIPv4Address {
+				addr = fmt.Sprintf("%s:%d", entry.AddrIPv4[0].String(), port)
+				m.addAppAddressIPv4(appID, addr)
+			}
+			if hasIPv6Address {
```

Comment:
here I can see you are using the same logic as was before ... but is it fair to assume that ipv6 has preference over ipv4 if both are there for an app?

Reply:
it's a fair point so please open a new issue to track that.

Reply:
done https://github.com/dapr/components-contrib/issues/1606

Reply:
thanks